File: survival-common/src/main/java/net/survival/block/Chunk.java
package net.survival.block;

import java.util.ArrayList;

import it.unimi.dsi.fastutil.objects.Object2ShortArrayMap;
import net.survival.util.XIntegerArray;

public class Chunk {
    public static final int XLENGTH = 16;
    public static final int YLENGTH = 16;
    public static final int ZLENGTH = 16;
    public static final int BASE_AREA = XLENGTH * ZLENGTH;
    public static final int VOLUME = BASE_AREA * YLENGTH;

    private XIntegerArray rawData;
    private ArrayList<Block> rawIdToBlockMap;
    private Object2ShortArrayMap<Block> blockToRawIdMap;

    public Chunk() {
        rawData = new XIntegerArray(VOLUME, 1);
        rawIdToBlockMap = new ArrayList<>(4);
        blockToRawIdMap = new Object2ShortArrayMap<>(4);
        rawIdToBlockMap.add(StandardBlocks.AIR);
        blockToRawIdMap.put(StandardBlocks.AIR, (short) 0);
    }

    public Chunk(XIntegerArray rawData, Block[] blockPalette) {
        this.rawData = rawData;
        rawIdToBlockMap = new ArrayList<>(blockPalette.length);
        blockToRawIdMap = new Object2ShortArrayMap<>(blockPalette.length);

        for (var i = 0; i < blockPalette.length; ++i) {
            rawIdToBlockMap.add(blockPalette[i]);
        }

        for (var i = 0; i < blockPalette.length; ++i) {
            blockToRawIdMap.put(blockPalette[i], (short) i);
        }
    }

    private Chunk(
            XIntegerArray rawData,
            ArrayList<Block> rawIdToBlockMap,
            Object2ShortArrayMap<Block> blockToRawIdMap)
    {
        this.rawData = rawData;
        this.rawIdToBlockMap = rawIdToBlockMap;
        this.blockToRawIdMap = blockToRawIdMap;
    }

    public Chunk makeCopy() {
        var copyOfRawData = rawData.makeCopy();
        var copyOfRawIdToBlockMap = new ArrayList<Block>(rawIdToBlockMap);
        var copyOfBlockToRawIdMap = new Object2ShortArrayMap<Block>(blockToRawIdMap);
        return new Chunk(copyOfRawData, copyOfRawIdToBlockMap, copyOfBlockToRawIdMap);
    }

    public XIntegerArray getRawData() {
        return rawData;
    }

    public Block[] getBlockPalette() {
        var array = new Block[rawIdToBlockMap.size()];
        rawIdToBlockMap.toArray(array);

        return array;
    }

    public Block getBlock(int index) {
        return rawIdToBlockMap.get((int) rawData.get(index));
    }

    public Block getBlock(int x, int y, int z) {
        return getBlock(localPositionToIndex(x, y, z));
    }

    public void setBlock(int index, Block to) {
        if (to == null) {
            throw new IllegalArgumentException("to");
        }

        if (blockToRawIdMap.containsKey(to)) {
            rawData.set(index, blockToRawIdMap.getShort(to));
        }
        else {
            var newRawId = blockToRawIdMap.size();
            blockToRawIdMap.put(to, (short) newRawId);
            rawIdToBlockMap.add(to);

            if (!rawData.isValidValue(newRawId)) {
                rawData = rawData.getResized(rawData.bitsPerElement + 1);

                // TODO: Perform garbage collection on the block palette.
            }

            rawData.set(index, newRawId);
        }
    }

    public void setBlock(int x, int y, int z, Block to) {
        var index = localPositionToIndex(x, y, z);
        setBlock(index, to);
    }

    public int localPositionToIndex(int x, int y, int z) {
        return x + (z * XLENGTH) + (y * BASE_AREA);
    }

    public boolean isInBounds(int x, int y, int z) {
        return x >= 0 && y >= 0 && z >= 0 && x < XLENGTH && y < YLENGTH && z < ZLENGTH;
    }
}

File: survival-common/src/main/java/net/survival/block/Chunk.kt
package net.survival.block

import it.unimi.dsi.fastutil.objects.Object2ShortArrayMap
import net.survival.util.XIntegerArray
import java.util.*

class Chunk {
    companion object {
        const val XLENGTH = 16
        const val YLENGTH = 16
        const val ZLENGTH = 16
        const val BASE_AREA = XLENGTH * ZLENGTH
        const val VOLUME = BASE_AREA * YLENGTH
    }

    var rawData: XIntegerArray; private set
    private val rawIdToBlockMap: ArrayList<Block>
    private val blockToRawIdMap: Object2ShortArrayMap<Block>

    constructor() {
        rawData = XIntegerArray(VOLUME, 1)
        rawIdToBlockMap = ArrayList(4)
        blockToRawIdMap = Object2ShortArrayMap(4)
        rawIdToBlockMap.add(StandardBlocks.AIR)
        blockToRawIdMap[StandardBlocks.AIR] = 0.toShort()
    }

    constructor(chunk: Chunk) {
        rawData = XIntegerArray(chunk.rawData)
        rawIdToBlockMap = ArrayList(chunk.rawIdToBlockMap)
        blockToRawIdMap = Object2ShortArrayMap(chunk.blockToRawIdMap)
    }

    constructor(rawData: XIntegerArray, blockPalette: Array<Block>) {
        this.rawData = rawData
        rawIdToBlockMap = ArrayList(blockPalette.size)
        blockToRawIdMap = Object2ShortArrayMap(blockPalette.size)
        for (i in blockPalette.indices) {
            rawIdToBlockMap.add(blockPalette[i])
        }
        for (i in blockPalette.indices) {
            blockToRawIdMap[blockPalette[i]] = i.toShort()
        }
    }

    val blockPalette: Array<Block> get() = rawIdToBlockMap.toTypedArray()

    fun getBlock(index: Int): Block {
        return rawIdToBlockMap[rawData[index].toInt()]
    }

    fun getBlock(x: Int, y: Int, z: Int): Block {
        return getBlock(localPositionToIndex(x, y, z))
    }

    fun setBlock(index: Int, block: Block) {
        if (blockToRawIdMap.containsKey(block)) {
            rawData[index] = blockToRawIdMap.getShort(block).toLong()
        }
        else {
            val newRawId = blockToRawIdMap.size
            blockToRawIdMap[block] = newRawId.toShort()
            rawIdToBlockMap.add(block)
            if (!rawData.isValidValue(newRawId.toLong())) {
                rawData = rawData.getResized(rawData.bitsPerElement + 1)

                // TODO: Perform garbage collection on the block palette.
            }
            rawData[index] = newRawId.toLong()
        }
    }

    fun setBlock(x: Int, y: Int, z: Int, block: Block) {
        val index = localPositionToIndex(x, y, z)
        setBlock(index, block)
    }

    fun localPositionToIndex(x: Int, y: Int, z: Int): Int =
        x + z * XLENGTH + y * BASE_AREA

    fun isInBounds(x: Int, y: Int, z: Int): Boolean =
        x >= 0 && y >= 0 && z >= 0 && x < XLENGTH && y < YLENGTH && z < ZLENGTH
}